File: leetcode-200_number-of-islands.go
/**
 * 给定一个由0和1组成的二维数组, 0表示海水, 1表示陆地, 求陆地的数量(连续的1算一个陆地)
 */
func numIslands(grid [][]byte) int {
    num := 0
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if string(grid[i][j]) == "1" {
                //如果当前坐标为陆地, 则将邻接当前坐标的陆地坐标都置0
                num++
                searchIsland(grid, i, j)
            }
        }
    }
    return num
}

func searchIsland(grid [][]byte, i int, j int) {
    //未超出边界并且为1则算陆地
    if i >= 0 && i < len(grid) && j >= 0 && j < len(grid[0]) && string(grid[i][j]) == "1" {
        grid[i][j] = '0' //将当前坐标的上下左右邻接陆地坐标都置0, 当作一整块陆地
        searchIsland(grid, i + 1, j)
        searchIsland(grid, i - 1, j)
        searchIsland(grid, i, j + 1)
        searchIsland(grid, i, j - 1)
    }
}

File: leetcode-200_number-of-islands.java
/**
 * 给定一个由'0'和'1'组成的二维数组, '0'表示海水, '1'表示陆地, 求陆地的数量(相邻的'1'算一个陆地)
 */
public int numIslands(char[][] grid) {
    int num = 0;
    for (int i = 0; i < grid.length; i++) {
        for (int j = 0; j < grid[0].length; j++) {
            // 遇到一块陆地
            if (grid[i][j] == '1') {
                num++;
                // 递归地将当前陆地与相邻陆地变成海水(即整块陆地记为一块)
                dfs(grid, i, j);
            }
        }
    }
    return num;
}

public void dfs(char[][] grid, int i, int j) {
    // 如果超出边界或已经是海水，无需处理
    if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] == '0') {
        return;
    }
    // 变成海水
    grid[i][j] = '0';
    // 继续处理相邻区域
    dfs(grid, i + 1, j);
    dfs(grid, i - 1, j);
    dfs(grid, i, j + 1);
    dfs(grid, i, j - 1);
}
